Add cuckoo-search package

rtrppl opened the pull request:
### Brief summary of what the package does

The package adds content-based search and saved-searches for the Elfeed package. Content-based search draws on `ripgrep`; saved-searches  are stored in a JSON-formatted local file.

### Direct link to the package repository

https://github.com/rtrppl/cuckoo-search

### Your association with the package

I am the maintainer of the package.

### Relevant communications with the upstream package maintainer

None needed.

### Checklist

<!-- Please confirm by replacing `[]` with `[x]`: -->

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] I've used `M-x checkdoc` to check the package's documentation strings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

<!-- After submitting, please fix any problems the CI reports. -->

riscy commented:
Thanks for this.  Quick first pass -

First, note we don't usually package the readme.org with the package.  I'd recommend removing that from the recipe, and making the package self-contained with its information (e.g. using your `;;; Commentary` and/or an occasional outside http link).

⸺ `cuckoo-search.el` with byte-compile using Emacs 30.1:
```
cuckoo-search.el:48:11: Error: Cannot open load file: No such file or directory, elfeed
```

⸺ `cuckoo-search.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- `;;; Commentary` should usually not redirect to README
- cuckoo-search.el#L120: Loading a package should not add advice
```

`#'load`-check on each file:
```
Loading cuckoo-search.el
  cuckoo-search.el:Error: Emacs 30.1:
  (file-missing "Cannot open load file" "No such file or directory" "elfeed")
```

⸺ Package and license:
- Please specify `:fetcher` before `:repo` in your recipe
- Add a LICENSE file to the repository

rtrppl commented:
Hi & thanks for your work and the feedback!

> First, note we don't usually package the readme.org with the package. I'd recommend removing that from the recipe, and making the package self-contained with its information (e.g. using your `;;; Commentary` and/or an occasional outside http link).

I removed the readme from the recipe.

> 
> ⸺ `cuckoo-search.el` with byte-compile using Emacs 30.1:
> 
> ```
> cuckoo-search.el:48:11: Error: Cannot open load file: No such file or directory, elfeed
> ```
> 

I have removed

```elisp
(require 'elfeed)
(require 'elfeed-db)
```

Another user of the package suggested that I should add these two, as I draw on the Elfeed package for certain of its functions. I noticed the same issue with byte-compile, though. 


> ⸺ `cuckoo-search.el` with [melpazoid](https://github.com/riscy/melpazoid):
> 
> ```
> - `;;; Commentary` should usually not redirect to README

I removed the reference to the README.

> - cuckoo-search.el#L120: Loading a package should not add advice
> ```

I needed to add this add-reference to "fix" the status line in Elfeed that I hack into. I'm happy to other ideas on how to fix this. 

> 
> `#'load`-check on each file:
> 
> ```
> Loading cuckoo-search.el
>   cuckoo-search.el:Error: Emacs 30.1:
>   (file-missing "Cannot open load file" "No such file or directory" "elfeed")
> ```

See above.

> ⸺ Package and license:
> 
> * Please specify `:fetcher` before `:repo` in your recipe

I changed the order.

> * Add a LICENSE file to the repository

I added a License file.

riscy commented:
> I needed to add this add-reference to "fix" the status line in Elfeed that I hack into.

I think this advice is now gone, but correct me if I'm wrong.  Here's the latest -

⸺ `cuckoo-search.el` with byte-compile using Emacs 30.1:
```
cuckoo-search.el:55:67: Warning: reference to free variable `elfeed-db-directory'
In cuckoo-search-get-index-meta:
cuckoo-search.el:81:29: Warning: Unused lexical argument `key'
In cuckoo-search:
cuckoo-search.el:108:23: Warning: reference to free variable `elfeed-search-entries'
cuckoo-search.el:111:14: Warning: assignment to free variable `elfeed-search-entries'
cuckoo-search.el:116:23: Warning: reference to free variable `elfeed-search-print-entry-function'
cuckoo-search.el:120:18: Warning: assignment to free variable `elfeed-search-last-update'
In cuckoo-search-saved-searches:
cuckoo-search.el:181:15: Warning: assignment to free variable `elfeed-search-filter'
In end of data:
cuckoo-search.el:182:10: Warning: the function `elfeed-search-update--force' is not known to be defined.
cuckoo-search.el:119:25: Warning: the function `elfeed-search--header' is not known to be defined.
cuckoo-search.el:82:46: Warning: the function `elfeed-entry-content' is not known to be defined.
```

⸺ `cuckoo-search.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- cuckoo-search.el#L122: Loading a package should not add advice
```

⸺ `cuckoo-search.el` with checkdoc 0.6.2 (fix *within reason*):
```
cuckoo-search.el:26: White space found at end of line
```

`#'load`-check on each file:
```
Loading cuckoo-search.el
  cuckoo-search.el:Error: Emacs 30.1:
  (void-variable elfeed-db-directory)
```

⸺ Package and license:
- Prefer default recipe: `(cuckoo-search :fetcher github :repo "rtrppl/cuckoo-search")`

rtrppl commented:
Hi Chris,

Thanks for the feedback. This seems to be Catch-22. All the warnings are related to functions that are within the Elfeed package (https://github.com/skeeto/elfeed/tree/master). What would be the correct way to prevent the warnings?

If I add 

```elisp
(require 'elfeed)
(require 'elfeed-db)
```

we are back to square one (see the error in your first response). 

My package is an extension for the Elfeed package. The author has long maintained that he is not interested in adding content-based search. This is a work-around for those users (including me) that want this feature. As such, Elfeed is of course still required.

The advice-add is still in the code. I hack into the `header-line-format` of Elfeed to add the Cuckoo-Search search terms there. When the user invokes `elfeed-search-clear-filter` the previous state is restored.

riscy commented:
I understand -

If you want to depend on elfeed and require users to install it, then you should add it to your `Package-Requires` line.  This seems to be what you want to do, because you're immediately referring to the variable `elfeed-db-directory` in your `cuckoo-search-elfeed-data-folder` defvar.  And `cuckoo-search-elfeed-data-folder` is explicitly referenced in `cuckoo-search` with no guards as to whether `elfeed` is installed or not.

If you want to make elfeed optional, you'll have to make your code robust to the case where `elfeed` is *not* installed, and you can use a combination of `defvar` and `declare-function` to pacify the byte-compiler, e.g.

```elisp
(defvar elfeed-db-directory)
(declare-function elfeed-search-update--force "ext:elfeed.el" nil t)
```

rtrppl commented:
Sorry for the late reply, been on the road the last few days. I've re-added the package-requires line, as this seems to be the most sensible option. Without Elfeed installed, my package does not make a lot of sense.

riscy commented:
Thank you!  I can see you need to add an elfeed version to your `;; Package-Requires` line.  Assuming that is there, the only issues remaining are:

⸺ `cuckoo-search.el` with byte-compile using Emacs 30.1:
```
cuckoo-search.el:83:29: Warning: Unused lexical argument `key'
```

⸺ `cuckoo-search.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- cuckoo-search.el#L124: Loading a package should not add advice
```
Ideally a package can be loaded into Emacs without making unexpected modifications to the runtime.  Advice is one of the big ones, because it's often used as a hack (and I realize cuckoo-search describes itself as a collection of hacks) or as a workaround when the upstream package doesn't have a hook defined.

I feel like the pattern that fits here is to define a `cuckoo-search-mode` that adds this advice (and turns it back off again) and then use `define-globalized-minor-mode` to define a `cuckoo-search-global-mode` that turns on your minor mode when `(derived-mode-p 'elfeed-search-mode)` .  Users would turn on `cuckoo-search-global-mode` once in their config, but they could also turn it off again to restore their runtime.

This is a lot - but I think you end up with an idiomatic Emacs package instead of a collection of hacks.  Let me know what you think...

⸺ `cuckoo-search.el` with checkdoc 0.6.2 (fix *within reason*):
```
cuckoo-search.el:26: White space found at end of line
```

Also since your recipe just tags `:defaults`, you can use this instead:

```elisp
(cuckoo-search :fetcher github :repo "rtrppl/cuckoo-search")
```

rtrppl commented:
Thanks!

> Thank you! I can see you need to add an elfeed version to your `;; Package-Requires` line. Assuming that is there, the only issues remaining are:

Sorry, I wonder how to add a version number to `(require 'elfeed)` or did I misunderstand this part? Elfeed is currently on version "3.4.2", but my package should also work with earlier versions. 

> ⸺ `cuckoo-search.el` with byte-compile using Emacs 30.1:
> 
> ```
> cuckoo-search.el:83:29: Warning: Unused lexical argument `key'
> ```

I've changed the line to:

`(maphash (lambda (_key value)` to indicate that key is not used. Completely dropping key does not work.

> Ideally a package can be loaded into Emacs without making unexpected modifications to the runtime. Advice is one of the big ones, because it's often used as a hack (and I realize cuckoo-search describes itself as a collection of hacks) or as a workaround when the upstream package doesn't have a hook defined.
> 
> I feel like the pattern that fits here is to define a `cuckoo-search-mode` that adds this advice (and turns it back off again) and then use `define-globalized-minor-mode` to define a `cuckoo-search-global-mode` that turns on your minor mode when `(derived-mode-p 'elfeed-search-mode)` . Users would turn on `cuckoo-search-global-mode` once in their config, but they could also turn it off again to restore their runtime.
> 
> This is a lot - but I think you end up with an idiomatic Emacs package instead of a collection of hacks. Let me know what you think...

I've added both modes. Would I be allowed to activate `cuckoo-search-global-mode` by adding (cuckoo-search-global-mode) to the end of the package? This would avoid breaking changes for existing users.

> Also since your recipe just tags `:defaults`, you can use this instead:
> 
> ```emacs-lisp
> (cuckoo-search :fetcher github :repo "rtrppl/cuckoo-search")
> ```

Will change this.

riscy commented:
> Sorry, I wonder how to add a version number to (require 'elfeed)

Not your `(require 'elfeed)` line, your `Package-Requires` line.  I.e. `;; Package-Requires: ((emacs "29.1")) (elfeed "3.4.2"))`

> but my package should also work with earlier versions.

Unless you want to test them, I'd just go with 3.4.2.  This signals to users that they'll get the expected behavior if they're at least on elfeed 3.4.2.

I think your code is still changing, so I'm seeing some other new issues, fyi -

⸺ `cuckoo-search.el` with byte-compile using Emacs 30.1:
```
cuckoo-search.el:71:31: Warning: in defcustom for `cuckoo-search-global-mode': fails to specify containing group
In cuckoo-search-global-mode:
cuckoo-search.el:71:2: Warning: docstring has wrong usage of unescaped single quotes (use \=' or different quoting such as `...')
```

⸺ `cuckoo-search.el` with package-lint 20250418.1425:
```
1 issue found:
71:0: error: Global minor modes should be autoloaded or, rarely, `:require' their defining file (i.e. ":require 'cuckoo-search"), to support the customization variable of the same name.
```

⸺ `cuckoo-search.el` with checkdoc 0.6.2 (fix *within reason*):
```
cuckoo-search.el:66: White space found at end of line
```

rtrppl commented:
Thanks again.

 > Not your `(require 'elfeed)` line, your `Package-Requires` line. I.e. `;; Package-Requires: ((emacs "29.1")) (elfeed "3.4.2"))`

That does make way more sense :). 


> I think your code is still changing, so I'm seeing some other new issues, fyi -
> 
> ⸺ `cuckoo-search.el` with byte-compile using Emacs 30.1:
> 
> ```
> cuckoo-search.el:71:31: Warning: in defcustom for `cuckoo-search-global-mode': fails to specify containing group
> In cuckoo-search-global-mode:
> cuckoo-search.el:71:2: Warning: docstring has wrong usage of unescaped single quotes (use \=' or different quoting such as `...')

I have now added a group and changed this whole part (see below). I have also set `:init-value t`, which appears to be much cleaner solution for loading `cuckoo-search-global-mode`.

```elisp
(defgroup cuckoo-search ()
  "Content-based search and saved-searches for Elfeed."
  :group 'comm)

(define-minor-mode cuckoo-search-mode
  "Minor mode to better integrate `cuckoo-search' into `elfeed-search-mode'."
  :lighter " cuckoo-search-mode"
  (if cuckoo-search-mode
      (advice-add 'elfeed-search-clear-filter :after #'cuckoo-search-elfeed-restore-header)
    (advice-remove 'elfeed-search-clear-filter #'cuckoo-search-elfeed-restore-header)))

;;;###autoload 
(define-globalized-minor-mode cuckoo-search-global-mode
  cuckoo-search-mode
  (lambda ()
    (when (derived-mode-p 'elfeed-search-mode)
      (cuckoo-search-mode 1)))
  :group 'cuckoo-search
  :init-value t)
```
> 
> ⸺ `cuckoo-search.el` with package-lint 20250418.1425:
> 
> ```
> 1 issue found:
> 71:0: error: Global minor modes should be autoloaded or, rarely, `:require' their defining file (i.e. ":require 'cuckoo-search"), to support the customization variable of the same name.
> ```

I've added an autoload, see above. Thanks again for taking the time to look into this.

riscy commented:
Thank you for the updates!  Here's the latest -

⸺ `cuckoo-search.el` with byte-compile using Emacs 30.1:
```
cuckoo-search.el:78:2: Warning: docstring has wrong usage of unescaped single quotes (use \=' or different quoting such as `...')
```
This is a funny one.  `define-globalized-minor-mode` is a macro, and its macroexpansion creates the following docstring:
```elisp
    "Toggle Cuckoo-Search mode in all buffers.
With prefix ARG, enable Cuckoo-Search-Global mode if ARG is positive; otherwise,
disable it.

If called from Lisp, toggle the mode if ARG is `toggle'.
Enable the mode if ARG is nil, omitted, or is a positive number.
Disable the mode if ARG is a negative number.

Cuckoo-Search mode is enabled in all buffers where `(lambda nil (when
(derived-mode-p 'elfeed-search-mode) (cuckoo-search-mode 1)))' would do it.

See `cuckoo-search-mode' for more information on Cuckoo-Search mode."
```
The byte-compiler is complaining about the single-quoting of `'elfeed-search-mode) (cuckoo-search-mode 1)))'`.  I don't fully understand why Emacs decided to make single-quoting a byte-compile *warning* recently, and this is also arguably a bug in Emacs, but you should not be using a lambda for this.  You can fix this by turning `(lambda () (when (derived-mode-p 'elfeed-search-mode) (cuckoo-search-mode 1)))` into a named function.

⸺ `cuckoo-search.el` with package-lint 20250418.1425:
```
1 issue found:
1:0: error: There is no (provide 'cuckoo-search) form.
```
Also I notice your code runs `(cuckoo-search-global-mode)` - don't do this!  Let users do this in their configs.

⸺ `cuckoo-search.el` with checkdoc 0.6.2 (fix *within reason*):
```
cuckoo-search.el:77: White space found at end of line
```

rtrppl commented:
Thanks for the feedback!

> The byte-compiler is complaining about the single-quoting of `'elfeed-search-mode) (cuckoo-search-mode 1)))'`. I don't fully understand why Emacs decided to make single-quoting a byte-compile _warning_ recently, and this is also arguably a bug in Emacs, but you should not be using a lambda for this. You can fix this by turning `(lambda () (when (derived-mode-p 'elfeed-search-mode) (cuckoo-search-mode 1)))` into a named function.

Done.

> ```
> 1 issue found:
> 1:0: error: There is no (provide 'cuckoo-search) form.
> ```
> 
> 
> Also I notice your code runs `(cuckoo-search-global-mode)` - don't do this!  Let users do this in their configs.

This was a mistake. I intended to remove `(cuckoo-search-global-mode)` but did remove `(provide 'cuckoo-search)`. Not sure how this happened. I've made my peace with adding this breaking change to the package and have adjusted the documentation accordingly.

If you don't mind, how do I run these more comprehensive tests myself? I seem to be unable to move beyond the complain 

> cuckoo-search.el:48:11: Error: Cannot open load file: No such file or directory, elfeed

riscy commented:
> If you don't mind, how do I run these more comprehensive tests myself?

Those are via [melpazoid](https://github.com/riscy/melpazoid) which is just my own bundle of scripts - it handles installing the dependencies in a new image.

This is minor but I'd recommend making your `:lighter` shorter, e.g. `cuckoo-search` instead of `cuckoo-search-mode` (`:lighter " cuckoo-search-mode"`).

Also I notice your minor mode (local to a buffer) has a global effect (`(advice-add 'elfeed-search-clear-filter :after #'cuckoo-search-elfeed-restore-header)`).  Probably you want to check if `cuckoo-search-mode` is active in `cuckoo-search-elfeed-restore-header` (although I'm not familiar with elfeed the local/global behavior of what you're trying to achieve).  Not a blocker though...

I'll merge this now!

rtrppl commented:
Great & thank you! I learned a lot in doing this and may soon come back with a much larger project. :) 

(I will take a look at :lighter and local/global behaviour.)
